presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectSelectiveStreamReader.java

```diff
@@ -0,0 +1,550 @@
+
+    @Override
+    public void close()
+    {
+        systemMemoryContext.close();
+    }
+
+    private void openRowGroup()
+            throws IOException
+    {
+        presentStream = presentStreamSource.openStream();
+        lengthStream = lengthStreamSource.openStream();
+        dataStream = dataStreamSource.openStream();
+
+        rowGroupOpen = true;
+    }
+
+    @Override
+    public void startStripe(InputStreamSources dictionaryStreamSources, List<ColumnEncoding> encoding)
+    {
+        presentStreamSource = missingStreamSource(BooleanInputStream.class);
+        lengthStreamSource = missingStreamSource(LongInputStream.class);
+        dataStreamSource = missingStreamSource(ByteArrayInputStream.class);
+
+        readOffset = 0;
+
+        presentStream = null;
+        lengthStream = null;
+        dataStream = null;
+
+        rowGroupOpen = false;
+    }
+
+    @Override
+    public void startRowGroup(InputStreamSources dataStreamSources)
+    {
+        presentStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, PRESENT, BooleanInputStream.class);
+        lengthStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, LENGTH, LongInputStream.class);
+        dataStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, DATA, ByteArrayInputStream.class);
+
+        readOffset = 0;
+
+        presentStream = null;
+        lengthStream = null;
+        dataStream = null;
+
+        rowGroupOpen = false;
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE + sizeOf(offsets) + sizeOf(outputPositions) + sizeOf(data) + sizeOf(nulls) + sizeOf(lengths) + sizeOf(isNullVector);
+    }
+
+    private void prepareForNextRead(int positionCount, int[] positions)
+            throws IOException
+    {
+        lengthIndex = 0;
+        outputPositionCount = 0;
+
+        int totalLength = 0;
+        int maxLength = 0;
+
+        int totalPositions = positions[positionCount - 1] + 1;
+        int nullCount = 0;
+        if (presentStream != null) {
+            isNullVector = ensureCapacity(isNullVector, totalPositions);
+            nullCount = presentStream.getUnsetBits(totalPositions, isNullVector);
+        }
+
+        if (lengthStream != null) {
+            lengths = ensureCapacity(lengths, totalPositions - nullCount);
+            lengthStream.nextIntVector(totalPositions - nullCount, lengths, 0);
+
+            offsets = ensureCapacity(offsets, totalPositions + 1);
+
+            for (int i = 0; i < totalPositions - nullCount; i++) {
+                totalLength += lengths[i];
+                maxLength = Math.max(maxLength, lengths[i]);
+            }
+
+            if (totalLength > ONE_GIGABYTE) {
+                throw new PrestoException(
+                        GENERIC_INTERNAL_ERROR,
+                        format("Values in column \"%s\" are too large to process for Presto. %s column values are larger than 1GB [%s]",
```

Comment:
a future improvement would be to add up lengths only for requested positions

Reply:
will takecare of this in a later commit, will add a TODO